src/NodeVisitor/NamespaceStmtCreator.php

```diff
@@ -0,0 +1,125 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of the humbug/php-scoper package.
+ *
+ * Copyright (c) 2017 Théo FIDRY <theo.fidry@gmail.com>,
+ *                    Pádraic Brady <padraic.brady@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Humbug\PhpScoper\NodeVisitor;
+
+use Humbug\PhpScoper\NodeVisitor\Collection\NamespaceStmtCollection;
+use PhpParser\Node;
+use PhpParser\Node\Stmt\Class_;
+use PhpParser\Node\Stmt\ClassLike;
+use PhpParser\Node\Stmt\Interface_;
+use PhpParser\Node\Stmt\Namespace_;
+use PhpParser\NodeVisitorAbstract;
+
+/**
+ * Creates a namespace for whitelisted classes and interfaces which belong to the global namespace.
+ *
+ * ```
+ * class AppKernel {}
+ * ```
+ *
+ * =>
+ *
+ * ```
+ * namespace Humbug;
+ *
+ * class AppKernel {}
+ * ```
+ */
+final class NamespaceStmtCreator extends NodeVisitorAbstract
+{
+    private $prefix;
+    private $hasWhitelistedNode;
+    private $namespaceStatements;
+    private $globalWhitelister;
+
+    /**
+     * @param string                  $prefix              Global prefix to apply.
+     * @param NamespaceStmtCollection $namespaceStatements List of statements in the current scope.
+     * @param callable                $globalWhitelister   List of whitelisted nodes.
+     */
+    public function __construct(
+        string $prefix,
+        NamespaceStmtCollection $namespaceStatements,
+        callable $globalWhitelister
+    ) {
+        $this->prefix = $prefix;
+        $this->namespaceStatements = $namespaceStatements;
+        $this->globalWhitelister = $globalWhitelister;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function beforeTraverse(array $nodes)
+    {
+        $classes = array_filter($nodes, [$this, 'isWhitelistableNode']);
+        if (empty($classes)) {
+            return;
+        }
+
+        $this->hasWhitelistedNode = $this->hasWhitelistedNode($classes);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function leaveNode(Node $node)
+    {
+        return $this->hasWhitelistedNode ? $this->wrapClassNamespace($node) : $node;
+    }
+
+    /**
+     * @return Node
+     */
+    private function wrapClassNamespace(Node $node): Node
+    {
+        if (AppendParentNode::hasParent($node)) {
+            return $node;
+        }
+
+        if (
+            $node instanceof ClassLike
+            && $this->isWhitelistableNode($node)
+            && ($this->globalWhitelister)($node->name)
+        ) {
+            return new Namespace_(new Node\Name($this->prefix), [$node]);
+        }
+
+        // Anything else needs to be wrapped with global namespace.
+        return new Namespace_(null, [$node]);
+    }
+
+    /**
+     * @param Node[] $nodes List of nodes to find whitelisted nodes from.
```

Comment:
I think the signature is descriptive enough, no need for the `List [...]` comment. I would keep the phpdoc as it does add the type `array`, but I tend to avoid comments when the code is self-explanatory as the problem with comments/docs is that they easily get out of date with the code they are documenting. So keeping it to minimal is better IMO

Reply:
Do you prefer to have an empty docblock or just none at all?

Reply:
In the projects we do at Yoast, we have very strict documentation rules.
We don't find ourselves having a lot of outdated documentation entries in general.

The fact that we are in a WordPress environment, where backwards compatibility is a major thing might have something to do with it, but I don't think that is the only factor in it.

Writing the documentation mostly helps me re-evaluate what I wrote functionally and if the naming of the variables and logic are actually the best solutions.

My two cents on the subject.

I'm happy to oblige to the standards and approach you set for your project.

Reply:
I think `@param Node[] $nodes` is enough here

If it's too "redundant" with the code, I really prefer to avoid the comments. For public interfaces, non trivial things or for example the doc to explain the purpose of the class I tend to be in favour comments :)